# Préprocess au NLP

Tout d’abord, install le paquet NLTK avec le gestionnaire de paquets + pip +:

In [1]:
#pip install nltk==3.3

Importation des librairies

In [2]:
import nltk
import pandas as pd
import numpy as np
from nltk.corpus import twitter_samples
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\ouizb\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

Téléchargez des commentaires Tripadvisor

In [3]:
df = pd.read_csv('resampled_comments_2.csv')
df

Unnamed: 0                                            comment  rating  \
0               1  Accueil sympa pas mal d’attente au final le ke...     3.0   
1               7  L’accueil est bonne mais les sandwichs mauvais...     1.0   
2               1  En balade dans le quartier nous cherchons un k...     3.0   
3               6  kebab identique à ailleurs cela reste un kebab...     3.0   
4               5  Au moment où je rédige cet avis le restaurant ...     1.0   
...           ...                                                ...     ...   
13305           5  Ce n'est pas difficile face à l'Elysee vous re...     5.0   
13306           7  Nous avons vraiment passé un excellent moment ...     5.0   
13307           3  Nous arrivons au service de 21h heure à laquel...     5.0   
13308           1  Ici comme à l'Eden Roc ou dans un autre hotel ...     5.0   
13309           2  Excellente pizzaria. Un accueil très chaleureu...     5.0   

                                              bonus_info   city sentiment  
0      Détails FOURCHETTE DE PRIX 2 895 AMD - 12 061 ...  Paris   Neutral  
1      Détails FOURCHETTE DE PRIX 2 895 AMD - 12 061 ...  Paris  Negative  
2      Détails FOURCHETTE DE PRIX 2 895 AMD - 12 061 ...  Paris   Neutral  
3      Détails FOURCHETTE DE PRIX 2 895 AMD - 12 061 ...  Paris   Neutral  
4      Détails FOURCHETTE DE PRIX 2 895 AMD - 12 061 ...  Paris  Negative  
...                                                  ...    ...       ...  
13305  Détails CUISINES Française Européenne RÉGIMES ...  Paris  Positive  
13306  Détails CUISINES Française Européenne RÉGIMES ...  Paris  Positive  
13307  Détails FOURCHETTE DE PRIX 2 628 AMD - 13 140 ...  Paris  Positive  
13308  Détails CUISINES Française Européenne RÉGIMES ...  Paris  Positive  
13309  Emplacement et coordonnées 43 Rue de Charenton...  Paris  Positive  

[13310 rows x 6 columns]

### On  a séparer les commentaires en fonction de leur note en 3 catégories:
- supérieur ou égale à 4 => positif
- égale à 3 => neutre
- inférieur ou égale à 2 => négatif

Je prends juste les commentaires, les notes  et les sentiments pour le moment.

In [4]:
comments = df[['comment', 'rating', 'sentiment']]
comments.head(10)

comment  rating sentiment
0  Accueil sympa pas mal d’attente au final le ke...     3.0   Neutral
1  L’accueil est bonne mais les sandwichs mauvais...     1.0  Negative
2  En balade dans le quartier nous cherchons un k...     3.0   Neutral
3  kebab identique à ailleurs cela reste un kebab...     3.0   Neutral
4  Au moment où je rédige cet avis le restaurant ...     1.0  Negative
5  Comment reconnaître des avis laissés par un pa...     1.0  Negative
6  Ce n’est pas possible que cet endroit soit cla...     1.0  Negative
7  Très surpris du nombre de profil avec 1 seul c...     1.0  Negative
8  Pas de salle juste deux tabourets a la limite ...     1.0  Negative
9  Je suis allée dans ce kebab après avoir vu les...     1.0  Negative

## Suppression du bruit

On met tout le texte en minuscule avnat la tokenization

In [5]:
comments['comment'] = comments['comment'].map(lambda x: x.lower())
comments

C:\Users\ouizb\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


comment  rating sentiment
0      accueil sympa pas mal d’attente au final le ke...     3.0   Neutral
1      l’accueil est bonne mais les sandwichs mauvais...     1.0  Negative
2      en balade dans le quartier nous cherchons un k...     3.0   Neutral
3      kebab identique à ailleurs cela reste un kebab...     3.0   Neutral
4      au moment où je rédige cet avis le restaurant ...     1.0  Negative
...                                                  ...     ...       ...
13305  ce n'est pas difficile face à l'elysee vous re...     5.0  Positive
13306  nous avons vraiment passé un excellent moment ...     5.0  Positive
13307  nous arrivons au service de 21h heure à laquel...     5.0  Positive
13308  ici comme à l'eden roc ou dans un autre hotel ...     5.0  Positive
13309  excellente pizzaria. un accueil très chaleureu...     5.0  Positive

[13310 rows x 3 columns]

## Tokenisation

Tokenization des phrases en mot.

In [6]:
from nltk import RegexpTokenizer
toknizer = RegexpTokenizer(r'''\w'|\w+|[^\w\s]''')
token = comments.apply(lambda row: toknizer.tokenize(row['comment']),axis= 1)
#token

Suppression de la ponctuation

In [7]:
import string
import re 
#string.punctuation
token_1 = token.apply(lambda x: [item.strip(string.punctuation) for item in x])
token_2 = token_1.apply(lambda x: [re.sub(r'\’', '', item) for item in x])
#token_2

suppression des nombres

In [8]:
token_4 = token_2.apply(lambda x: [re.sub(r'\d', '', item) for item in x])

#print(token_4) 

## Normalisation

### Stopword Removal : Suppression des mots vides

In [9]:
from nltk.corpus import stopwords 
import nltk 
nltk.download('stopwords')
stop_words = set(stopwords.words('french')) 
stop_token = token_4.apply(lambda x: [item for item in x if item not in stop_words])

print(stop_token)

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\ouizb\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
0        [accueil, sympa, mal, , attente, final, kebab,...
1        [, accueil, bonne, sandwichs, mauvais, plus, t...
2        [balade, quartier, cherchons, kebab, voici, tr...
3        [kebab, identique, ailleurs, cela, reste, keba...
4        [moment, où, rédige, cet, avis, restaurant, cl...
                               ...                        
13305    [difficile, face, elysee, remontez, rue, côté,...
13306    [vraiment, passé, excellent, moment, nouveau, ...
13307    [arrivons, service, h, heure, laquelle, réserv...
13308    [ici, comme, eden, roc, autre, hotel, groupe, ...
13309    [excellente, pizzaria, , accueil, très, chaleu...
Length: 13310, dtype: object


### Stemming

In [10]:
from nltk.stem.snowball import SnowballStemmer
stemmer = SnowballStemmer(language='french')
stemm = stop_token.apply(lambda x: [stemmer.stem(y) for y in x])

print(stemm)

0        [accueil, symp, mal, , attent, final, kebab, c...
1        [, accueil, bon, sandwich, mauv, plus, tromp, ...
2        [balad, quarti, cherchon, kebab, voic, tripadv...
3        [kebab, ident, ailleur, cel, rest, kebab, tran...
4        [moment, où, rédig, cet, avis, restaur, class,...
                               ...                        
13305    [difficil, fac, elyse, remont, ru, côt, bristo...
13306    [vrai, pass, excellent, moment, nouveau, resta...
13307    [arrivon, servic, h, heur, laquel, réserv, dîn...
13308    [ici, comm, eden, roc, autr, hotel, group, lev...
13309    [excellent, pizzari, , accueil, tres, chaleur,...
Length: 13310, dtype: object


### Lemmatization

In [11]:
#pip install git+https://github.com/ClaudeCoulombe/FrenchLefffLemmatizer.git

In [12]:
from french_lefff_lemmatizer.french_lefff_lemmatizer import FrenchLefffLemmatizer
lemmatizer = FrenchLefffLemmatizer()
lemm = stemm.apply(lambda x: [lemmatizer.lemmatize(y) for y in x])
lemm

0        [accueil, symp, mal, , attent, final, kebab, c...
1        [, accueil, bon, sandwich, mauv, plus, tromp, ...
2        [balad, quarti, cherchon, kebab, voic, tripadv...
3        [kebab, ident, ailleur, cel, rest, kebab, tran...
4        [moment, où, rédig, cet, avis, restaur, class,...
                               ...                        
13305    [difficil, fac, elyse, remont, ru, côt, bristo...
13306    [vrai, pass, excellent, moment, nouveau, resta...
13307    [arrivon, servic, heure, heur, laquel, réserv,...
13308    [ici, comm, eden, roc, autr, hotel, group, lev...
13309    [excellent, pizzari, , accueil, tres, chaleur,...
Length: 13310, dtype: object

Maintenant, assemblons les tokens traités ensemble.

In [13]:
for i in range(len(lemm)):
    lemm[i] = ' '.join(lemm[i])


In [14]:
# On rajoute à notre dataframe les données lemmatiser comme nouvelle colonne
comments['lemmatiser_com'] = lemm
comments

C:\Users\ouizb\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


comment  rating sentiment  \
0      accueil sympa pas mal d’attente au final le ke...     3.0   Neutral   
1      l’accueil est bonne mais les sandwichs mauvais...     1.0  Negative   
2      en balade dans le quartier nous cherchons un k...     3.0   Neutral   
3      kebab identique à ailleurs cela reste un kebab...     3.0   Neutral   
4      au moment où je rédige cet avis le restaurant ...     1.0  Negative   
...                                                  ...     ...       ...   
13305  ce n'est pas difficile face à l'elysee vous re...     5.0  Positive   
13306  nous avons vraiment passé un excellent moment ...     5.0  Positive   
13307  nous arrivons au service de 21h heure à laquel...     5.0  Positive   
13308  ici comme à l'eden roc ou dans un autre hotel ...     5.0  Positive   
13309  excellente pizzaria. un accueil très chaleureu...     5.0  Positive   

                                          lemmatiser_com  
0      accueil symp mal  attent final kebab correct s...  
1       accueil bon sandwich mauv plus tromp command ...  
2      balad quarti cherchon kebab voic tripadvisor i...  
3      kebab ident ailleur cel rest kebab transcend  ...  
4      moment où rédig cet avis restaur class eme tri...  
...                                                  ...  
13305  difficil fac elyse remont ru côt bristol  seco...  
13306  vrai pass excellent moment nouveau restaur esp...  
13307  arrivon servic heure heur laquel réserv dîn  a...  
13308  ici comm eden roc autr hotel group levur chimi...  
13309  excellent pizzari  accueil tres chaleur pizz v...  

[13310 rows x 4 columns]

###  Fonctionnalités TF-IDF
Cette méthode prend en compte non seulement l'occurrence d'un mot dans un seul commentaire (ou tweet) mais aussi dans le corpus entier.
TF-IDF fonctionne en pénalisant les mots communs en leur attribuant des poids inférieurs tout en donnant de l'importance aux mots qui sont rares dans l'ensemble du corpus mais qui apparaissent en bon nombre dans peu de commentaires.

Termes importants liés à TF-IDF:
- TF = (Nombre de fois que le terme t apparaît dans un document) / (Nombre de termes dans le document)
- IDF = log (N / n), où, N est le nombre de documents et n est le nombre de documents dans lesquels un terme t est apparu.
- TF-IDF = TF * IDF

In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer(max_df = 0.90 , min_df = 2 , max_features = 1000 , stop_words = stop_words, ngram_range=(1, 2)) 
# TF-IDF matrice de fonctionnalités
X = tfidf_vectorizer.fit_transform(comments['lemmatiser_com'])


### Fonctionnalités Bag-of-Words

In [16]:
from sklearn.feature_extraction.text import CountVectorizer
bow_vectorizer = CountVectorizer(max_df=0.90, min_df=2, max_features=1000, stop_words= stop_words)
# bag-of-words matrice de fonctionnalités
X_bow = bow_vectorizer.fit_transform(comments['lemmatiser_com'])

## Construction de modèles: analyse des sentiments

In [17]:
y = comments['sentiment']

#### Split trani et test set

In [18]:
from sklearn.model_selection import train_test_split
# TF-IDF matrice
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)
# bag-of-words matrice
X_bow_train, X_bow_test, y_train, y_test = train_test_split(X_bow, y, test_size=0.2, random_state=1)

### Random Forest

In [19]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score

# TF-IDF matrice
rf = RandomForestClassifier()
rf.fit(X_train, y_train)
scores = cross_val_score(rf, X_train, y_train, cv=5)
print(scores, "Avg accuracy Random Forest avec TF-IDF matrice: {}".format(np.mean(scores)))

[0.80328638 0.80140845 0.8028169  0.79426961 0.80648192] Avg accuracy Random Forest avec TF-IDF matrice: 0.8016526527254964


In [20]:
# bag-of-words matrice
rf_2 = RandomForestClassifier()
rf_2.fit(X_bow_train, y_train)
scores_2 = cross_val_score(rf_2, X_bow_train, y_train, cv=5)
print(scores_2, "Avg accuracy Random Forest avec bag-of-words matrice: {}".format(np.mean(scores_2)))

[0.79295775 0.80938967 0.81032864 0.79567872 0.80366369] Avg accuracy Random Forest avec bag-of-words matrice: 0.8024036941234065


### Naive bayes

####  Naive bayes multinomial

In [21]:
from sklearn.naive_bayes import MultinomialNB
# TF-IDF matrice
mnb = MultinomialNB()
mnb.fit(X_train.todense(), y_train)
scores_mnb = cross_val_score(mnb, X_train, y_train, cv=5)
print(scores_mnb, "Avg accuracy MultinomialNB avec TF-IDF matrice: {}".format(np.mean(scores_mnb)))

[0.70234742 0.70938967 0.71173709 0.70737435 0.7007985 ] Avg accuracy MultinomialNB avec TF-IDF matrice: 0.706329405901512


In [22]:
# bag-of-words matrice
mnb2 = MultinomialNB()
mnb2.fit(X_bow_train.todense(), y_train)
scores_mnb2 = cross_val_score(mnb2, X_bow_train.todense(), y_train, cv=5)
print(scores_mnb2, "Avg accuracy MultinomialNB avec bag-of-words matrice: {}".format(np.mean(scores_mnb2)))

[0.71549296 0.71596244 0.70798122 0.70784406 0.70925317] Avg accuracy MultinomialNB avec bag-of-words matrice: 0.71130676969284


### SVM

In [46]:
from sklearn.svm import SVC
# TF-IDF matrice
svc_model = SVC(gamma='auto')
svc_model.fit(X_train, y_train)
scores_svm = cross_val_score(svc_model, X_train, y_train, cv=5)
print(scores_svm, "Avg accuracy SVM avec TF-IDF matrice: {}".format(np.mean(scores_svm)))

[0.38920188 0.38920188 0.38873239 0.38891498 0.38938469] Avg accuracy SVM avec TF-IDF matrice: 0.38908716428837625


In [24]:
# bag-of-words matrice
#svc_model2 = SVC(gamma='auto')
#svc_model2.fit(X_bow_train, y_train)
#scores_svm2 = cross_val_score(svc_model2, X_bow_train, y_train, cv=5)
#print(scores_svm2, "Avg accuracy SVM avec bag-of-words matrice: {}".format(np.mean(scores_svm2)))

### Logistic régression

In [25]:
from sklearn.linear_model import LogisticRegression
# TF-IDF matrice
logreg = LogisticRegression()
logreg.fit(X_train, y_train)
scores_lg = cross_val_score(logreg, X_train,y_train , cv=5)
print(scores_lg, "Avg accuracy LogisticRegression: {}".format(np.mean(scores_lg)))
#warnings.filterwarnings("ignore")

C:\Users\ouizb\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\ouizb\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-reg

[0.72535211 0.72910798 0.73802817 0.7242837  0.7350869 ] Avg accuracy LogisticRegression: 0.7303717718869975


In [27]:
#bag-of-words matrice
logreg2 = LogisticRegression()
logreg2.fit(X_bow_train, y_train)
scores_lg2 = cross_val_score(logreg2, X_bow_train, y_train, cv=5)
print(scores_lg2, "Avg accuracy LogisticRegression: {}".format(np.mean(scores_lg2)))
#warnings.filterwarnings("ignore")

C:\Users\ouizb\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\ouizb\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-reg

[0.73521127 0.72910798 0.72394366 0.71911696 0.72804133] Avg accuracy LogisticRegression: 0.7270842402150495


C:\Users\ouizb\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


Random Forest à l'accuracy la plus élevée avec TF-IDF matrice avec 80.16%.

Random Forest à l'accuracy la plus élevée avec bag-of-words matrice avec 80.24%.

# On teste nos prédictions

In [28]:
y_pred = rf_2.predict(X_bow_test) # RF avec bag-of-words matrice

In [29]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

print(confusion_matrix(y_test,y_pred))
print(classification_report(y_test,y_pred))
print(accuracy_score(y_test, y_pred))

[[427 138  58]
 [ 45 935  98]
 [ 18 119 824]]
              precision    recall  f1-score   support

    Negative       0.87      0.69      0.77       623
     Neutral       0.78      0.87      0.82      1078
    Positive       0.84      0.86      0.85       961

    accuracy                           0.82      2662
   macro avg       0.83      0.80      0.81      2662
weighted avg       0.83      0.82      0.82      2662

0.8211870773854245


### Grid search cv

LR

In [ ]:
from sklearn.model_selection import GridSearchCV
warnings.filterwarnings("ignore")
# Create regularization penalty space
penalty = ['l1', 'l2']

# Create regularization hyperparameter space
C = np.logspace(0, 4, 10)

# Create hyperparameter options
hyperparameters = dict(C=C, penalty=penalty)

# Create grid search using 5-fold cross validation
clf = GridSearchCV(logreg, hyperparameters, cv=5, verbose=0)

# Fit grid search
best_model_lr = clf.fit(X, y)

# View best hyperparameters
print('Best Penalty:', best_model_lr.best_estimator_.get_params()['penalty'])
print('Best C:', best_model_lr.best_estimator_.get_params()['C'])

# Predict target vector
best_model_lr.predict(X)

RF

In [30]:
from sklearn.model_selection import GridSearchCV
parameters = {
    'n_estimators'      : [0.001, 0.01, 0.1,1,10,100, 100],
    'max_depth'         : [ 60, 70, 80, 90, 100],
    'random_state'      : [0],
    #'max_features': ['auto'],
    #'criterion' :['gini']
}

clf = GridSearchCV(rf_2, parameters, cv=10, n_jobs=-1)
clf.fit(X_bow_train, y_train)

print(clf.score(X_bow, y))
print(clf.best_params_)
#print(classification_report(y_test, clf.predict(X_bow_test), digits=4))
#print(zip(correct_train[predictor].columns))

0.9627347858752817
{'max_depth': 80, 'n_estimators': 100, 'random_state': 0}


NBM

In [31]:
parameters = {'alpha': [0.001, 0.01, 0.1, 1, 10, 100]}
clf_nbm = GridSearchCV(mnb2, parameters, cv=10)
clf_nbm.fit(X_bow_train, y_train)
print(clf_nbm.score(X_bow, y))
print(clf_nbm.best_params_)
print(classification_report(y_test, clf_nbm.predict(X_bow_test), digits=4))

0.7393688955672427
{'alpha': 0.01}
              precision    recall  f1-score   support

    Negative     0.6050    0.5778    0.5911       623
     Neutral     0.6733    0.6939    0.6834      1078
    Positive     0.8253    0.8210    0.8232       961

    accuracy                         0.7126      2662
   macro avg     0.7012    0.6976    0.6992      2662
weighted avg     0.7122    0.7126    0.7123      2662



Teste des nouveaux parametres

In [32]:
import pickle
# bag-of-words matrice
rf_3 = RandomForestClassifier(max_depth = 80, n_estimators = 100, random_state = 0)
rf_3.fit(X_bow_train, y_train)

# save the model to disk
#filename = 'final_model_malika.sav'
#pickle.dump(rf_3, open(filename, 'wb'))

scores_3 = cross_val_score(rf_3, X_bow_train, y_train, cv=5)
print(scores_3, "Avg accuracy Random Forest avec bag-of-words matrice: {}".format(np.mean(scores_3)))
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
y_pred = rf_3.predict(X_bow_test) # RF avec bag-of-words matrice
print(confusion_matrix(y_test,y_pred))
print(classification_report(y_test,y_pred))
print(accuracy_score(y_test, y_pred))

# load the model from disk
#loaded_model = pickle.load(open(filename, 'rb'))
#result = loaded_model.score(X_bow_test, y_test)
#print(result)

[0.7971831  0.80610329 0.80985915 0.79755754 0.80178488] Avg accuracy Random Forest avec bag-of-words matrice: 0.8024975908370215
[[411 162  50]
 [ 46 933  99]
 [ 21 115 825]]
              precision    recall  f1-score   support

    Negative       0.86      0.66      0.75       623
     Neutral       0.77      0.87      0.82      1078
    Positive       0.85      0.86      0.85       961

    accuracy                           0.81      2662
   macro avg       0.83      0.79      0.80      2662
weighted avg       0.82      0.81      0.81      2662

0.8148009015777611


In [34]:
# bag-of-words matrice
mnb3 = MultinomialNB(alpha = 0.01)
mnb3.fit(X_bow_train.todense(), y_train)
scores_mnb3 = cross_val_score(mnb3, X_bow_train.todense(), y_train, cv=5)
print(scores_mnb3, "Avg accuracy MultinomialNB avec bag-of-words matrice: {}".format(np.mean(scores_mnb2)))
y_pred2 = mnb3.predict(X_bow_test) # NBM avec bag-of-words matrice
print(confusion_matrix(y_test,y_pred2))
print(classification_report(y_test,y_pred2))
print(accuracy_score(y_test, y_pred2))

[0.71455399 0.71549296 0.7084507  0.70878347 0.71395021] Avg accuracy MultinomialNB avec bag-of-words matrice: 0.71130676969284
[[360 227  36]
 [199 748 131]
 [ 36 136 789]]
              precision    recall  f1-score   support

    Negative       0.61      0.58      0.59       623
     Neutral       0.67      0.69      0.68      1078
    Positive       0.83      0.82      0.82       961

    accuracy                           0.71      2662
   macro avg       0.70      0.70      0.70      2662
weighted avg       0.71      0.71      0.71      2662

0.7126220886551465


## Ensemble learning

### Bagging

In [39]:
from sklearn.ensemble import BaggingClassifier
from sklearn import tree
from sklearn.metrics import accuracy_score, classification_report, precision_score, recall_score 
from sklearn.metrics import confusion_matrix, precision_recall_curve, roc_curve, auc, log_loss
from sklearn.metrics import accuracy_score
model_bag = BaggingClassifier(tree.DecisionTreeClassifier(random_state=1))
model_bag.fit(X_train, y_train)
model_bag.score(X_test,y_test)

y_pred_bag = model_bag.predict(X_test)


print(accuracy_score(y_test, y_pred_bag))


cnf_matrix_bag = confusion_matrix(y_test, y_pred_bag)
print(cnf_matrix_bag)

0.7595792637114951
[[463 125  35]
 [104 877  97]
 [ 81 198 682]]


In [40]:
model_bag2 = BaggingClassifier(tree.DecisionTreeClassifier(random_state=1))
model_bag2.fit(X_bow_train, y_train)
model_bag2.score(X_bow_test,y_test)

y_pred_bag2 = model_bag2.predict(X_bow_test)


print(accuracy_score(y_test, y_pred_bag2))


cnf_matrix_bag2 = confusion_matrix(y_test, y_pred_bag2)
print(cnf_matrix_bag2)

0.7573253193087904
[[442 134  47]
 [102 873 103]
 [ 70 190 701]]


### Ada boost

In [41]:
from sklearn.ensemble import AdaBoostClassifier

model_ada = AdaBoostClassifier(random_state=1)
model_ada.fit(X_train, y_train)
#model.score(x_test,y_test)

y_pred_ada = model_ada.predict(X_test)

print(accuracy_score(y_test, y_pred_ada))


cnf_matrix_ada = confusion_matrix(y_test, y_pred_ada)
print(cnf_matrix_ada)

0.6036814425244177
[[285 255  83]
 [263 630 185]
 [ 80 189 692]]


In [42]:
model_ada2 = AdaBoostClassifier(random_state=1)
model_ada2.fit(X_train, y_train)
#model.score(x_test,y_test)

y_pred_ada2 = model_ada.predict(X_bow_test)

print(accuracy_score(y_test, y_pred_ada2))


cnf_matrix_ada2 = confusion_matrix(y_test, y_pred_ada2)
print(cnf_matrix_ada2)

0.36626596543951917
[[ 88 355 180]
 [109 620 349]
 [113 581 267]]


### Stacking

In [61]:
from sklearn.ensemble import StackingClassifier

estimators = [('Random forest', rf), ("Naive bayes", mnb),("SVM", svc_model)]
reg = StackingClassifier(estimators=estimators, final_estimator=RandomForestClassifier())
reg.fit(X_train, y_train)
scores_reg = reg.score(X_train, y_train)

print(scores_reg, "Avg accuracy RandomForestClassifier: {}".format(np.mean(scores_reg)))

0.9941773102930128 Avg accuracy RandomForestClassifier: 0.9941773102930128


In [62]:
y_pred_reg = reg.predict(X_test)
print(accuracy_score(y_test, y_pred_reg))

0.8395942900075132


In [63]:
cnf_matrix_reg = confusion_matrix(y_test, y_pred_reg)
print(cnf_matrix_reg)

[[488 110  25]
 [101 911  66]
 [ 32  93 836]]


In [70]:
estimators = [('Random forest', rf), ("Naive bayes", mnb),("SVM", svc_model)]
reg2 = StackingClassifier(estimators=estimators, final_estimator=RandomForestClassifier())
reg2.fit(X_bow_train, y_train)
scores_reg2 = reg2.score(X_bow_train, y_train)
accu_reg2 = np.mean(scores_reg2)
print(scores_reg2, "Avg accuracy RandomForestClassifier: {}".format(accu_reg2))

0.9973703981968445 Avg accuracy RandomForestClassifier: 0.9973703981968445


In [71]:
y_pred_reg2 = reg.predict(X_bow_test)
print(accuracy_score(y_test, y_pred_reg2))

0.37340345604808417


In [68]:
estimators = [('Random forest', rf), ("Naive bayes", mnb),("SVM", svc_model)]
reg3 = StackingClassifier(estimators=estimators, final_estimator=RandomForestClassifier(max_depth = 80, n_estimators = 100, random_state = 0))
reg3.fit(X_train, y_train)
scores_reg3 = reg3.score(X_train, y_train)

print(scores_reg3, "Avg accuracy RandomForestClassifier: {}".format(np.mean(scores_reg3)))

0.9929564237415477 Avg accuracy RandomForestClassifier: 0.9929564237415477


In [69]:
y_pred_reg3 = reg3.predict(X_test)
print(accuracy_score(y_test, y_pred_reg3))

0.8373403456048084


In [ ]:
estimators = [('Random forest', rf), ("Naive bayes", mnb),("SVM", svc_model)]
reg3 = StackingClassifier(estimators=estimators, final_estimator=RandomForestClassifier(max_depth = 80, n_estimators = 100, random_state = 0))
reg3.fit(X_train, y_train)
scores_reg3 = reg3.score(X_train, y_train)

print(scores_reg3, "Avg accuracy RandomForestClassifier: {}".format(np.mean(scores_reg3)))

In [ ]:
y_pred_reg3 = reg3.predict(X_test)
print(accuracy_score(y_test, y_pred_reg3))

### Voting

In [53]:
from sklearn.ensemble import VotingClassifier
voting = VotingClassifier(estimators=[('SVM', svc_model), ('Naive bayes', mnb),('Random forest', rf) ], voting='hard')
voting.fit(X_train, y_train)
scores_voting = cross_val_score(voting, X_train, y_train, cv=5)
print(scores_voting, "Avg accuracy Voting: {}".format(np.mean(scores_voting)))


[0.73004695 0.74553991 0.74976526 0.73790512 0.73790512] Avg accuracy Voting: 0.7402324704450279


In [54]:
y_pred_voting = voting.predict(X_test)
print(accuracy_score(y_test, y_pred_voting))

0.7426746806912096
